Documentation/google-slave-provisioning.md

```diff
@@ -0,0 +1,265 @@
+cat > jenkins-agent.json <<EOF
+{
+    "builders": [
+        {
+            "type": "googlecompute",
+            "project_id": "$PROJECT",
+            "source_image_family": "ubuntu-1604-lts",
+            "source_image_project_id": "ubuntu-os-cloud",
+            "zone": "europe-west2-b",
+            "disk_size": "20",
+            "image_name": "jenkins-agent-{{timestamp}}",
+            "image_family": "jenkins-agent",
+            "ssh_username": "ubuntu"
+        }
+    ],
+    "provisioners": [
+        {
+            "type": "shell",
+            "inline": [
+                "sudo apt-get update",
+                "sudo apt-get install -y default-jdk"
+            ]
+        }
+    ]
+}
+EOF
+```
+*However*; since you are not able to provide `user-data` and `init-script` with Jenkins at a later stage, this is where you'll need to add these scripts. To do this we will need to add additional [provisioners](https://www.packer.io/docs/provisioners/index.html). In short, there are two types of provisioners of interest: [shell](https://www.packer.io/docs/provisioners/shell.html) which runs the specified script(s) (can also be inline) **once** when packer builds the script. This is where you want to place your `user-data` script. The other provisioner you'll need is [file](https://www.packer.io/docs/provisioners/file.html), which adds a file from your local file system to the image, typically you have to add the file to some destination, and use shell to move it to where you want it if it requires sudo. 
+
+Now, typically you will want (part of) your `init-script` to run when the agent boots. *(You should put as much of your scripts as possible in the shell provisioner to save compute time and money.)* To make this happen, you will have to make a little work-around. One way to do this, is to provide your *script* as a *file*, and add it to `/etc/init.d/`, and finally use a *shell* provisioner to make it execute in `rc.local`. For example:
+```
+export PROJECT=$(gcloud info --format='value(config.project)')
+cat > jenkins-agent.json <<EOF
+{
+    "builders": [
+        {
+            "type": "googlecompute",
+            "project_id": "jenkins-195709",
+            "source_image_family": "ubuntu-1604-lts",
+            "source_image_project_id": "ubuntu-os-cloud",
+            "zone": "europe-west2-b",
+            "disk_size": "10",
+            "image_name": "jenkins-agent-{{timestamp}}",
+            "image_family": "jenkins-agent",
+            "ssh_username": "ubuntu"
+        }
+    ],
+    "provisioners": [
+        {
+            "type": "shell",
+            "execute_command": "echo 'packer' | sudo -S sh -c '{{ .Vars }} {{ .Path }}'",
+            "scripts": [
+                "/path/to/user_data.sh",
+                "/path/to/non-init/init_script.sh"
+            ]
+        },
+        {
+                "type": "file",
+                "source": "/path/to/init/init-script.sh",
```

Comment:
I think I know the answer, but just to verify: You need to upload this script to the packer host first, right? It doesn't grab it from mender-qa?

I don't see a reference to mender-qa, so I assume so.

Reply:
No, this is just a script local on your cloud console. But you probably can pull it directly from github by doing an inline shell provisioner as follows:
{
    "type": "shell",
    "inline": [ 
        "sudo apt update", 
        "sudo apt install -y git", 
        "git clone https://github.com/mender-qa", 
        "bash /path/to/script/"
 ]
}